superduperdb/base/leaf.py

```diff
@@ -1,15 +1,73 @@
+import dataclasses as dc
+import importlib
 import inspect
 import typing as t
 from abc import ABC, abstractmethod, abstractproperty
+from superduperdb.misc.serialization import asdict
 
 _CLASS_REGISTRY = {}
 
+if t.TYPE_CHECKING:
+    from superduperdb.base.datalayer import Datalayer
 
+
+def _import_item(cls, module, dict, db: t.Optional['Datalayer'] = None):
+    module = importlib.import_module(module)
+    cls = getattr(module, cls)
+    try:
+        return cls(**dict, db=db)
+    except TypeError as e:
+        if 'got an unexpected keyword argument' in str(e):
+            if callable(cls):
+                return cls(
+                    **{k: v for k, v in dict.items() if k in inspect.signature(cls).parameters},
+                    db=db
+                )
+            return cls(
+                **{k: v for k, v in dict.items() if k in inspect.signature(cls.__init__).parameters},
+                db=db
+            )
+        raise e
+
+
+@dc.dataclass
 class Leaf(ABC):
+    db: dc.InitVar[t.Optional['Datalayer']] = None
+
+    def __post_init__(self, db):
+        self.db: 'Datalayer' = db
+
     def __init_subclass__(cls, **kwargs):
         super().__init_subclass__(**kwargs)
         cls._register_class()
 
+    def encode(self):
+        cache = {}
+        blobs = {}
+        files = {}
+        self._deep_flat_encode(cache, blobs, files, ())
+        return {
+            '_base': f'?{self.id}',
+            '_leaves': cache,
+        }
+
+    def _deep_flat_encode(self, cache, blobs, files, leaves_to_keep=()):
+        if isinstance(self, leaves_to_keep):
+            cache[self.id] = self
+            return f'?{self.id}'
+        from superduperdb.base.document import _deep_flat_encode
```

Comment:
I think we should refrain from using private function outside the class .. 
if it is intended to be used outside the class we can remove `_`